### ISSUE TITLE
Unusual Unicode character in sender name messes up UI alignment

### ISSUE THREAD

**Issue by christian-heusel** [neomutt/neomutt#4460]
## NeoMutt Version

```
neomutt -v
NeoMutt 20241114
Copyright (C) 2015-2024 Richard Russon and friends
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 6.12.0-1-mainline (x86_64)
ncurses: ncurses 6.5.20240427 (compiled with 6.5.20240427)
libidn2: 2.3.7 (compiled with 2.3.7)
GPGME: 1.24.0
GnuTLS: 3.8.8
libnotmuch: 5.6.0
storage: kyotocabinet, gdbm, lmdb
compression: lz4, zlib, zstd

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --sqlite --autocrypt --lua --notmuch --gss --gnutls --sasl --with-ui=ncurses --with-idn2=/usr --disable-idn --idn2 --lmdb --kyotocabinet --gdbm --ubsan --lz4 --zlib --zstd

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection         -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -g -ffile-prefix-map=/build/neomutt/src=/usr/src/debug/neomutt -flto=auto -std=c11 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -D_XOPEN_SOURCE_EXTENDED -I/usr/include/lua5.3 -I/usr/include -I/usr/include -I/usr/include -DNCURSES_WIDECHAR -I/usr/include -I/usr/include/p11-kit-1 -I/usr/include -I/usr/include -fsanitize=undefined -fno-omit-frame-pointer -fno-common -g -O0

Compile options:
  +autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo +gnutls +gpgme 
  -gsasl +gss +hcache -homespool +idn +inotify -locales_hack +lua +nls +notmuch 
  -openssl +pgp +regex +sasl +smime +sqlite +truecolor 

Devel options:
  ubsan 

MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

## Expected Behaviour

Lines are self-contained

## Actual Behaviour

For a certain user the line is overrun and the tree view of the mails is not aligned. This is similar to what ia have previously reported in #4443, but has a different cause.
One can see the overrun in the box to the left under the "inbox" name and the displacement in the index format.

![image](https://github.com/user-attachments/assets/288f4c9f-48f5-4735-b359-d298f4e24984)

## Steps to Reproduce

```
set index_format="%4C %Z %?X?A& ? %D %-15.15F %s (%-4.4c)"
```

Have someone with Unicode Character 12288 in the Name send you a mail (can be obtained by doing `python -c 'print(f">{chr(12288)}<")'` for example).

## How often does this happen?

- [ ] Always

- [ ] Sometimes

- [x] On a particular email: https://pkgbuild.com/~gromit/file.eml
  If the email isn't private, please attach it to this issue.

## When did it start to happen?

- [ ] When I upgraded
  Which version did you use to use?

- [ ] When I changed my config
  Can you narrow down **what** you changed?

- [x] When I first got email from said user

## Extra Info

* Operating System and its version: Arch Linux (rolling)
* Were you using multiple copies of NeoMutt at once?: no
* Were you using 'screen' or 'tmux'?: no
* Is your email local (maildir) or remote (IMAP)?: local

**Comment by flatcap** [comment]
Ah, sorry!
That _may_ have already been fixed, in 62db1ef5b
Formatted expandos at the end of a format string weren't being truncated correctly.

If you're happy to build NeoMutt, please can you try the `[main]` branch.
(There's a lot of development going on, but it should be stable enough :-)

Thanks!

**Comment by christian-heusel** [comment]
The issue is not fixed on the main branch, I had my local version already patched for the issue that I had reported in #4443.

**Comment by flatcap** [comment]
ok, I'll have a closer look in the morning

**Comment by christian-heusel** [comment]
Thanks a ton! 🤗 I'm happy to provide further debugging info if needed, I have also linked a sample `.eml` file above.

**Comment by gahr** [comment]
TL;DR - I don't think there's much we can do, unless we switch to using some other library for UTF-8.

Right now we rely on the host's libc, and I think your libc (and mine too, FWIW) doesn't recognize U+12288 as being a wide character...

See what's the output of compiling and running this sample

```c
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <wchar.h>

int main()
{
  setlocale(LC_ALL, "");

  const char utf8[] = "\xF0\x92\x8A\x88";
  wchar_t wc;
  size_t n = mbstowcs(&wc, utf8, 1);
  int w = wcwidth(wc);

  printf("Converted %zu chars\n", n);
  printf("Its width is %d\n", w);
  printf("|12345|\n");

  // printing the utf-8 bytes
  printf("|%s|\n", utf8);

  // printing the wide char
  putchar('|');
  putwchar(wc);
  putchar('|');
  putchar('\n');
}
```

with this command line: `LC_ALL=en_US.UTF-8 cc -o wcwidth wcwidth.c && ./wcwidth`. On my system, as you can see, it wrongly identifies the char width as being 1, and the second `|` character on the line ends up under `2`...

![Screenshot 2024-12-02 at 08 59 53](https://github.com/user-attachments/assets/c264e550-a48e-4e8a-80e7-e625bd59d9bf)

Here on markdown, you can see that the correct output would be

```
|12345|
|𒊈|
|𒊈|
```

**Comment by gahr** [comment]
Ouch - I messed up.. I tested unicode char U+12288 , while the email you linked contains U+3000.

**Comment by christian-heusel** [comment]
Yeah sorry to lure you on the wrong path with my report, I have wrongly assumed that `ord()` in python would give me the unicode characters ID:
> Given a string representing one Unicode character, return an integer representing the Unicode code point of that character. For example, `ord('a')` returns the integer 97 and `ord('\u2020')` returns 8224. This is the inverse of [`chr()`](https://docs.python.org/3.4/library/functions.html#chr).

https://docs.python.org/3.4/library/functions.html#ord

I did the following:
```
>>> s = "Stefan 　 "
>>> for i in s:
...     print(ord(i))
83
116
101
102
97
110
32
12288
32
```

Since I already fear that we would miscommunicate about this I had also included a way to get it into my report, but I definitely should have been more clear ... 😬 
> can be obtained by doing `python -c 'print(f">{chr(12288)}<")'`
